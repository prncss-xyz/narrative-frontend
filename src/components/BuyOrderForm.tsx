import { Country } from "../hooks/countries";
import { Dataset } from "../hooks/datasets";
import { setPresenceInArray } from "../utils/arrays";
import {
  availableRecordCountForDatasets,
  forcastedRecordCount,
} from "../utils/logic";
import { convertMoney, normalizeMoney } from "../utils/money";
import { Box, Flex, Grid, H3 } from "./basics";
import { DatasetItemSmall } from "./DataItemSmall";
import { Input } from "./Input";
import { ToggleButton } from "./ToggleButton";

export type BuyOrderContents = {
  id?: string;
  name: string;
  createdAt?: Date;
  datasetIds: number[];
  countries: string[];
  budget: number;
};

/**
 * The common component to view/edit a buyOrder
 * buOrder.id and buyOrder.createdAt will be undefined to create a new buyOrder
 * setBuyOrder wil be undefined for viewing (readonly)
 * child component is used to place buttons a the bottom of the form
 */
export function BuyOrderForm({
  buyOrder,
  setBuyOrder,
  datasets,
  countries,
  children,
}: {
  buyOrder: BuyOrderContents;
  setBuyOrder?: (BuyOrder: BuyOrderContents) => void;
  datasets: Dataset[];
  countries: Country[];
  children: React.ReactNode;
}) {
  const forcasted = forcastedRecordCount(countries, datasets, buyOrder);
  const available = availableRecordCountForDatasets(
    countries,
    datasets,
    buyOrder.countries
  );
  const disabled = setBuyOrder === undefined;
  return (
    <Flex justifyContent="center">
      <Flex
        width={6}
        gap={3}
        flexDirection="column"
        px={4}
        py={3}
        backgroundColor="bgMid"
      >
        <Grid
          gridTemplateRows={"1fr 1fr"}
          gridTemplateColumns={"1fr 1fr"}
          gridRowGap={3}
          gridColumnGap={3}
        >
          <Box>
            <H3>Order name</H3>
            <Input
              py="0px"
              px={setBuyOrder ? 0 : 1}
              width="100%"
              borderStyle="none"
              maxHeight={1}
              placeholder="name"
              value={buyOrder.name}
              converter={String}
              setValue={
                setBuyOrder
                  ? (name) => setBuyOrder({ ...buyOrder, name })
                  : undefined
              }
            />
          </Box>
          <Box>
            {buyOrder.createdAt ? (
              <>
                <H3>Date Created</H3>
                <Box height={1} p="0px" alignItems="baseline">
                  {buyOrder.createdAt.toLocaleString([], {
                    year: "numeric",
                    month: "numeric",
                    day: "numeric",
                  })}
                </Box>
              </>
            ) : (
              <Box />
            )}
          </Box>
          <Box>
            <H3>Order budget</H3>
            <Flex alignItems="baseline">
              <Box>$</Box>
              <Input
                borderStyle="none"
                width="100%"
                maxHeight={1}
                py="0px"
                px={setBuyOrder ? 0 : 1}
                ml={setBuyOrder ? 2 : 0}
                placeholder="budget"
                value={buyOrder.budget}
                normalizer={normalizeMoney}
                converter={convertMoney}
                setValue={
                  setBuyOrder
                    ? (budget) => setBuyOrder({ ...buyOrder, budget })
                    : undefined
                }
              />
            </Flex>
          </Box>
          <Box>
            <H3>Forecasted Records</H3>
            <Box height={1} p="0px" alignItems="baseline">
              {`${forcasted} of ${available} available records`}
            </Box>
          </Box>
        </Grid>
        <Box>
          <H3>Included datasets</H3>
          <Grid
            gridTemplateColumns={"1fr 1fr"}
            alignItems="start"
            gridRowGap={3}
            gridColumnGap={3}
          >
            {datasets.map((dataset) => (
              <div key={dataset.id}>
                <DatasetItemSmall
                  dataset={dataset}
                  // To make this scalable, we would need to introduce
                  // the kind of abstraction provided by
                  // https://github.com/akheron/optics-ts
                  active={buyOrder.datasetIds.includes(dataset.id)}
                  setActive={
                    disabled
                      ? undefined
                      : (status) =>
                          setBuyOrder({
                            ...buyOrder,
                            datasetIds: setPresenceInArray(
                              buyOrder.datasetIds,
                              dataset.id,
                              status
                            ),
                          })
                  }
                />
              </div>
            ))}
          </Grid>
        </Box>
        <Box>
          <H3>Included countries</H3>
          <Flex gap={3}>
            {countries.map((country) => (
              <div key={country.countryCode}>
                <ToggleButton
                  active={buyOrder.countries.includes(country.countryCode)}
                  setActive={(status) =>
                    setBuyOrder
                      ? setBuyOrder({
                          ...buyOrder,
                          countries: setPresenceInArray(
                            buyOrder.countries,
                            country.countryCode,
                            status
                          ),
                        })
                      : undefined
                  }
                >
                  {country.name}
                </ToggleButton>
              </div>
            ))}
          </Flex>
        </Box>
        <Box mt={5} />
        {children}
      </Flex>
    </Flex>
  );
}
