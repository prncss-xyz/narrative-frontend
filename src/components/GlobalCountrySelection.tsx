import { createContext, useContext, useState } from "react";
import { Country } from "../hooks/countries";
import { setInArray } from "../utils/arrays";
import { Box, Flex, H3 } from "./basics";
import { ToggleButton } from "./ToggleButton";
import { TogglingSelector } from "./TogglingSelector";

type CountryCodes = string[];

const Context = createContext<
  null | [null | CountryCodes, (state: CountryCodes) => void]
>(null);

export function countryString(
  countries: Country[],
  activeCountryCodes: string[]
) {
  const activeCountryNames = countries
    .filter(({ countryCode }) => activeCountryCodes.includes(countryCode))
    .map(({ name }) => name)
    .sort();
  if (activeCountryNames.length === 0) return "none";
  if (activeCountryNames.length === 1) return activeCountryNames[0];
  const last = activeCountryNames.at(-1);
  const rest = activeCountryNames.slice(0, -1);
  return [rest.join(", "), last].join(" & ");
}

export function CountrySelectorContext({
  children,
}: {
  children: React.ReactNode;
}) {
  const [activeCountryCodes, setActiveCountryCodes] = useState<null | string[]>(
    null
  );
  return (
    <Context.Provider value={[activeCountryCodes, setActiveCountryCodes]}>
      {children}
    </Context.Provider>
  );
}

export function useGlobalCountyList(
  countries: Country[]
): [CountryCodes, (state: CountryCodes) => void] {
  const context = useContext(Context);
  if (!context)
    throw new Error(
      "CountrySelector should always be used inside CountrySelectorContext"
    );
  // eslint-disable-next-line prefer-const
  let [activeCountryCodes, setActiveCountryCodes] = context;
  activeCountryCodes ??= countries.map((country) => country.countryCode);
  return [activeCountryCodes, setActiveCountryCodes];
}

export function sortCountriesByName(countries: Country[]) {
  return [...countries].sort((a, b) =>
    a.name < b.name ? -1 : a.name > b.name ? 1 : 0
  );
}

export function GlobalCountrySelector({
  countries,
  ...props
}: {
  countries: Country[];
  [prop: string]: unknown; // TODO: could be more restrictive
}) {
  countries = sortCountriesByName(countries);
  const [activeCountryCodes, setActiveCountryCodes] =
    useGlobalCountyList(countries);
  return (
    <Box>
      <H3>Included countries</H3>
      <Flex gap={3} {...props}>
        {countries.map((country) => (
          <div key={country.countryCode}>
            <ToggleButton
              active={activeCountryCodes.includes(country.countryCode)}
              setActive={(status) =>
                // status
                setActiveCountryCodes(
                  // activeCountryCodes
                  setInArray(activeCountryCodes, country.countryCode, status)
                )
              }
            >
              {country.name}
            </ToggleButton>
          </div>
        ))}
      </Flex>
    </Box>
  );
}

export function GlobalCountrySelectionSummary({
  count,
  countries,
  ...props
}: {
  count: number;
  countries: Country[];
  [prop: string]: unknown; // TODO: could be more restrictive
}) {
  countries = sortCountriesByName(countries);
  const [activeCountryCodes, setActiveCountryCodes] =
    useGlobalCountyList(countries);
  return (
    <Flex
      justifyItems="start"
      flexWrap="wrap"
      color="muted"
      alignItems="baseline"
      pb={2}
      textAlign="left"
      columnGap="0.4em"
      {...props}
    >
      <Box display="inline">Showing</Box>
      <Box display="inline" color="text" fontWeight="bold">
        {count}
      </Box>
      <Box display="inline">{count === 1 ? "result" : "results"} from</Box>
      <Box display="inline" color="text" fontWeight="bold">
        {countryString(countries, activeCountryCodes)}
      </Box>
    </Flex>
  );
}
