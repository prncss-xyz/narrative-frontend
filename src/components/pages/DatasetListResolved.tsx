import { useState } from "react";
import { Country } from "../../hooks/countries";
import { Dataset } from "../../hooks/datasets";
import {
  availableRecordCountForDataset,
  includedCountries,
  selectedDatasets,
} from "../../utils/logic";
import { Box, Flex, Grid, H2, H3, Img } from "../basics";
import { Clickable } from "../Clickable";
import {
  countryString,
  GlobalCountrySelectionSummary,
  GlobalCountrySelector,
  useGlobalCountyList,
} from "../GlobalCountrySelection";
import { Overlay } from "../Overlay";

function DatasetItem({
  dataset,
  countries,
}: {
  dataset: Dataset;
  countries: Country[];
}) {
  const [activeCountryCodes] = useGlobalCountyList(countries);
  const available = availableRecordCountForDataset(
    countries,
    dataset,
    activeCountryCodes
  );
  return (
    <Flex
      p={3}
      gap={2}
      width={5}
      minHeight={5}
      flexDirection="column"
      backgroundColor="bgDark"
    >
      <Flex flexDirection="row" gap={2}>
        <Box maxHeight={3} minWidth={3} backgroundColor="bgLight">
          <Img
            src={dataset.thumbnailUrl}
            width={3}
            height={3}
            alt="dataset thumbnail"
          />
        </Box>
        <Flex alignItems="center" width="100%">
          <H2 my={0} width="100%">
            {dataset.label}
          </H2>
        </Flex>
      </Flex>
      <Flex flexDirection="column">
        <H3>Dataset Description</H3>
        <Box fontSize={1}>{dataset.description}</Box>
      </Flex>
      <Flex flexDirection="column" justifyContent="space-between">
        <H3 my={0}>Included countries</H3>
        <Box fontSize={1}>
          {countryString(
            countries,
            includedCountries(countries, dataset).map(
              (country) => country.countryCode
            )
          )}
        </Box>
      </Flex>
      <Flex
        flexDirection="row"
        justifyContent="space-between"
        alignItems="flex-end"
      >
        <H3 my={0}>Cost Per Record</H3>
        <Box>${dataset.costPerRecord}</Box>
      </Flex>
      <Flex flexDirection="row" justifyContent="space-between">
        <H3 my={0}>Available Records</H3>
        <Box>{`${available} records`}</Box>
      </Flex>
    </Flex>
  );
}

export function DatasetListResolved({
  datasets,
  countries,
}: {
  datasets: Dataset[];
  countries: Country[];
}) {
  const [activeCountryCodes] = useGlobalCountyList(countries);
  const selected = selectedDatasets(countries, datasets, activeCountryCodes);
  const [overlayVisible, setOverlayvisible] = useState(false);
  return (
    <>
      <Overlay
        visible={overlayVisible}
        setVisible={() => setOverlayvisible(false)}
      >
        <GlobalCountrySelector countries={countries} />
      </Overlay>
      <Flex flexDirection="row" justifyContent="center">
        <Box>
          <Clickable onClick={() => setOverlayvisible(true)}>
            <GlobalCountrySelectionSummary
              count={selected.length}
              countries={countries}
            />
          </Clickable>
          <Grid
            gridTemplateColumns={"auto auto"}
            alignItems="start"
            gridRowGap={4}
            gridColumnGap={4}
          >
            {selected.map((dataset) => (
              <div key={dataset.id}>
                <DatasetItem dataset={dataset} countries={countries} />
              </div>
            ))}
          </Grid>
        </Box>
      </Flex>
    </>
  );
}
