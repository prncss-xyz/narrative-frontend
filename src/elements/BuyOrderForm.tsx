import { useState } from "react";
import { BuyOrder } from "../hooks/buyOrders";
import { Country } from "../hooks/countries";
import { Dataset } from "../hooks/datasets";
import {
  availableRecordCountForDatasets,
  forcastedRecordCount,
} from "../utils/logic";
import { Box, Flex, Grid, H3 } from "./basics";
import { DatasetItemSmall } from "./DataItemSmall";
import { useGlobalCountyList } from "./GlobalCountrySelection";
import { Input } from "./Input";
import { RoundedButton } from "./RoundedButton";
import { TogglingSelector } from "./TogglingSelector";

function validateMoney(value: string) {
  if (value.match(/\..../)) return false; // no more than 2 decimals
  if (Number(value) >= 0) return true; // a number that is positve
  return false;
}

// export interface BuyOrderFormResult {
export type BuyOrderFormResult = {
  id?: string;
  name: string;
  createdAt?: Date;
  datasetIds: number[];
  countries: string[];
  budget: number;
};

export function BuyOrderForm({
  disabled,
  buyOrder,
  datasets,
  countries,
  toActions,
}: {
  disabled?: boolean;
  buyOrder?: BuyOrder;
  datasets: Dataset[];
  countries: Country[];
  toActions: (result: BuyOrderFormResult) => React.ReactNode;
}) {
  const [activeDatasets, setActiveDatasets] = useState<number[]>(
    buyOrder?.datasetIds ?? []
  );
  const [globalActiveCountries] = useGlobalCountyList(countries);
  const [activeCountryCodes, setActiveCountryCodes] = useState<string[]>(
    buyOrder?.countries ?? globalActiveCountries
  );
  const [name, setName] = useState(buyOrder?.name ?? "");
  const [budget, setBudget] = useState(String(buyOrder?.budget ?? 0));
  const result: BuyOrderFormResult = {
    ...buyOrder,
    name,
    budget: Number(budget),
    datasetIds: activeDatasets,
    countries: activeCountryCodes,
  };
  const forcasted = forcastedRecordCount(countries, datasets, result);
  const available = availableRecordCountForDatasets(
    countries,
    datasets,
    result.countries
  );

  return (
    <Flex justifyContent="center">
      <Flex flexDirection="column" width={6}>
        <Box px={4} py={3} backgroundColor="tone2">
          <Grid
            gridTemplateColumns={"1fr 1fr"}
            alignItems="start"
            gridRowGap={2}
            gridColumnGap={2}
          >
            <Box>
              <H3>Order name</H3>
              <Input
                disabled={disabled}
                validate={() => true}
                placeholder="name"
                value={name}
                setValue={setName}
              />
            </Box>
            <Box>
              {buyOrder && (
                <>
                  <H3>Date Created</H3>
                  {buyOrder.createdAt.toLocaleString([], {
                    year: "numeric",
                    month: "numeric",
                    day: "numeric",
                  })}
                </>
              )}
            </Box>
            <Box>
              <H3>Order budget</H3>
              $&nbsp;
              <Input
                disabled={disabled}
                validate={validateMoney}
                placeholder="budget"
                value={budget}
                setValue={setBudget}
              />
            </Box>
            <Box>
              <H3>Forecasted Records</H3>
              {`${forcasted} of ${available} available records`}
            </Box>
          </Grid>
          <Box mt={4} />
          <H3>Included datasets</H3>
          <Grid
            gridTemplateColumns={"1fr 1fr"}
            alignItems="start"
            gridRowGap={2}
            gridColumnGap={2}
          >
            <TogglingSelector
              disabled={disabled}
              state={activeDatasets}
              setState={setActiveDatasets}
              items={datasets.map((dataset) => ({
                key: dataset.id,
                toElem: (props) => (
                  <DatasetItemSmall dataset={dataset} {...props} />
                ),
              }))}
            />
          </Grid>
          <Box mt={4} />
          <H3>Included countries</H3>
          <Flex gap={3}>
            <TogglingSelector
              disabled={disabled}
              state={activeCountryCodes}
              setState={setActiveCountryCodes}
              items={countries.map((country) => ({
                key: country.countryCode,
                toElem: (props) => (
                  <RoundedButton
                    {...props}
                    // eslint-disable-next-line react/prop-types
                    borderStyle={props.active ? "solid" : "none"}
                  >
                    {country.name}
                  </RoundedButton>
                ),
              }))}
            />
          </Flex>
          <Box mt={6} />
          {toActions(result)}
        </Box>
      </Flex>
    </Flex>
  );
}
