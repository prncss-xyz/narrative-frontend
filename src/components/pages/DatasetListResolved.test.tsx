import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render } from "@testing-library/react";
import { BrowserRouter as Router, Route, Routes } from "react-router-dom";
import { FormBuyOrder } from "../BuyOrderForm";
import { CountrySelectorContext } from "../GlobalCountrySelection";
import { DatasetListResolved } from "./DatasetListResolved";

export const countries = [
  {
    countryCode: "US",
    name: "United States",
    storedData: [
      { datasetId: 1, recordCount: 500 },
      { datasetId: 2, recordCount: 0 },
      { datasetId: 3, recordCount: 2500 },
      { datasetId: 4, recordCount: 750 },
      { datasetId: 5, recordCount: 50 },
      { datasetId: 6, recordCount: 1000 },
    ],
  },
  {
    countryCode: "GB",
    name: "United Kingdom",
    storedData: [
      { datasetId: 1, recordCount: 0 },
      { datasetId: 2, recordCount: 1250 },
      { datasetId: 3, recordCount: 2000 },
      { datasetId: 4, recordCount: 1500 },
      { datasetId: 5, recordCount: 500 },
      { datasetId: 6, recordCount: 500 },
    ],
  },
];

export const datasets = [
  {
    id: 1,
    name: "fishing",
    label: "Fishing Zones",
    description:
      "Data about global fishing zones, including yields, historical trends and ocean life surveys.",
    thumbnailUrl: "https://picsum.photos/id/1038/92/92",
    costPerRecord: 0.03,
  },
  {
    id: 2,
    name: "shipping",
    label: "Shipping Routes",
    description:
      "Metadata regarding global shipping routes. Contains data such as traffic volumne, operating restrictiona and mileage of routes.",
    thumbnailUrl: "https://picsum.photos/id/1002/92/92",
    costPerRecord: 0.05,
  },
  {
    id: 3,
    name: "ports",
    label: "Ports of Entry",
    description: "Detailed information about ports related to cargo shipping.",
    thumbnailUrl: "https://picsum.photos/id/1026/92/92",
    costPerRecord: 0.02,
  },
  {
    id: 4,
    name: "ocean",
    label: "Ocean Measurements",
    description:
      "Detailed measurements of many ocean locations, including LAT/LON, temperature, humidity and precipitation.",
    thumbnailUrl: "https://picsum.photos/id/1024/92/92",
    costPerRecord: 0.01,
  },
  {
    id: 5,
    name: "weather",
    label: "Weather and Climate",
    description: "Historical averages and trends of oceanographic climate.",
    thumbnailUrl: "https://picsum.photos/id/135/92/92",
    costPerRecord: 0.04,
  },
];

describe("BuyOrderForm", () => {
  it("should render correctly", () => {
    const queryClient = new QueryClient();
    const { container } = render(
      <QueryClientProvider client={queryClient}>
        <CountrySelectorContext>
          <Router>
            <Routes>
              <Route
                path="/"
                element={
                  <DatasetListResolved
                    countries={countries}
                    datasets={datasets}
                  />
                }
              ></Route>
            </Routes>
          </Router>
        </CountrySelectorContext>
      </QueryClientProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
