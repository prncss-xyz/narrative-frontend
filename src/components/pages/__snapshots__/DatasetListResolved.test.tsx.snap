// Vitest Snapshot v1

exports[`BuyOrderForm > should render correctly 1`] = `
<div>
  <div
    class="css-1g12luf-Box-Flex e2i8k6n2"
  >
    <div
      class="css-1x3sxzl-Box e2i8k6n3"
    >
      <div
        class="css-1x3sxzl-Box e2i8k6n3"
      >
        <button
          class="css-19fpo8m-Box e2i8k6n3"
          cursor="pointer"
          height="100%"
        >
          <div
            class="css-104t6km-Box-Flex e2i8k6n2"
            color="muted"
          >
            <div
              class="css-4mhvi5-Box e2i8k6n3"
              display="inline"
            >
              Showing
            </div>
            <div
              class="css-1x0dezy-Box e2i8k6n3"
              color="text"
              display="inline"
              font-weight="bold"
            >
              5
            </div>
            <div
              class="css-4mhvi5-Box e2i8k6n3"
              display="inline"
            >
              results
               from
            </div>
            <div
              class="css-1x0dezy-Box e2i8k6n3"
              color="text"
              display="inline"
              font-weight="bold"
            >
              United Kingdom & United States
            </div>
          </div>
        </button>
      </div>
      <div
        class="css-1vhj0ch-Box-Grid e2i8k6n1"
      >
        <div>
          <div
            class="css-1y0qilw-Box-Flex e2i8k6n2"
            width="5"
          >
            <div
              class="css-wxi0h7-Box-Flex e2i8k6n2"
            >
              <div
                class="css-z7haim-Box e2i8k6n3"
              >
                <img
                  alt="dataset thumbnail"
                  class="css-1v0ofof-Img e2i8k6n4"
                  height="3"
                  src="https://picsum.photos/id/1038/92/92"
                  width="3"
                />
              </div>
              <div
                class="css-1hg49r7-Box-Flex e2i8k6n2"
                width="100%"
              >
                <h2
                  class="css-x4oqgn-Box e2i8k6n3"
                  font-size="4"
                  font-weight="normal"
                  width="100%"
                >
                  Fishing Zones
                </h2>
              </div>
            </div>
            <div
              class="css-d9aoz6-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-5jl3ge-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Dataset Description
              </h3>
              <div
                class="css-10zvdr1-Box e2i8k6n3"
                font-size="1"
              >
                Data about global fishing zones, including yields, historical trends and ocean life surveys.
              </div>
            </div>
            <div
              class="css-1rdjp9o-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-1mp3576-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Included countries
              </h3>
              <div
                class="css-10zvdr1-Box e2i8k6n3"
                font-size="1"
              >
                United States
              </div>
            </div>
            <div
              class="css-14cjmhn-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-1mp3576-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Cost Per Record
              </h3>
              <div
                class="css-1x3sxzl-Box e2i8k6n3"
              >
                $
                0.03
              </div>
            </div>
            <div
              class="css-1ezxmef-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-1mp3576-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Available Records
              </h3>
              <div
                class="css-1x3sxzl-Box e2i8k6n3"
              >
                500 records
              </div>
            </div>
          </div>
        </div>
        <div>
          <div
            class="css-1y0qilw-Box-Flex e2i8k6n2"
            width="5"
          >
            <div
              class="css-wxi0h7-Box-Flex e2i8k6n2"
            >
              <div
                class="css-z7haim-Box e2i8k6n3"
              >
                <img
                  alt="dataset thumbnail"
                  class="css-1v0ofof-Img e2i8k6n4"
                  height="3"
                  src="https://picsum.photos/id/1002/92/92"
                  width="3"
                />
              </div>
              <div
                class="css-1hg49r7-Box-Flex e2i8k6n2"
                width="100%"
              >
                <h2
                  class="css-x4oqgn-Box e2i8k6n3"
                  font-size="4"
                  font-weight="normal"
                  width="100%"
                >
                  Shipping Routes
                </h2>
              </div>
            </div>
            <div
              class="css-d9aoz6-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-5jl3ge-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Dataset Description
              </h3>
              <div
                class="css-10zvdr1-Box e2i8k6n3"
                font-size="1"
              >
                Metadata regarding global shipping routes. Contains data such as traffic volumne, operating restrictiona and mileage of routes.
              </div>
            </div>
            <div
              class="css-1rdjp9o-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-1mp3576-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Included countries
              </h3>
              <div
                class="css-10zvdr1-Box e2i8k6n3"
                font-size="1"
              >
                United Kingdom
              </div>
            </div>
            <div
              class="css-14cjmhn-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-1mp3576-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Cost Per Record
              </h3>
              <div
                class="css-1x3sxzl-Box e2i8k6n3"
              >
                $
                0.05
              </div>
            </div>
            <div
              class="css-1ezxmef-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-1mp3576-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Available Records
              </h3>
              <div
                class="css-1x3sxzl-Box e2i8k6n3"
              >
                1250 records
              </div>
            </div>
          </div>
        </div>
        <div>
          <div
            class="css-1y0qilw-Box-Flex e2i8k6n2"
            width="5"
          >
            <div
              class="css-wxi0h7-Box-Flex e2i8k6n2"
            >
              <div
                class="css-z7haim-Box e2i8k6n3"
              >
                <img
                  alt="dataset thumbnail"
                  class="css-1v0ofof-Img e2i8k6n4"
                  height="3"
                  src="https://picsum.photos/id/1026/92/92"
                  width="3"
                />
              </div>
              <div
                class="css-1hg49r7-Box-Flex e2i8k6n2"
                width="100%"
              >
                <h2
                  class="css-x4oqgn-Box e2i8k6n3"
                  font-size="4"
                  font-weight="normal"
                  width="100%"
                >
                  Ports of Entry
                </h2>
              </div>
            </div>
            <div
              class="css-d9aoz6-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-5jl3ge-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Dataset Description
              </h3>
              <div
                class="css-10zvdr1-Box e2i8k6n3"
                font-size="1"
              >
                Detailed information about ports related to cargo shipping.
              </div>
            </div>
            <div
              class="css-1rdjp9o-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-1mp3576-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Included countries
              </h3>
              <div
                class="css-10zvdr1-Box e2i8k6n3"
                font-size="1"
              >
                United Kingdom & United States
              </div>
            </div>
            <div
              class="css-14cjmhn-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-1mp3576-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Cost Per Record
              </h3>
              <div
                class="css-1x3sxzl-Box e2i8k6n3"
              >
                $
                0.02
              </div>
            </div>
            <div
              class="css-1ezxmef-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-1mp3576-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Available Records
              </h3>
              <div
                class="css-1x3sxzl-Box e2i8k6n3"
              >
                4500 records
              </div>
            </div>
          </div>
        </div>
        <div>
          <div
            class="css-1y0qilw-Box-Flex e2i8k6n2"
            width="5"
          >
            <div
              class="css-wxi0h7-Box-Flex e2i8k6n2"
            >
              <div
                class="css-z7haim-Box e2i8k6n3"
              >
                <img
                  alt="dataset thumbnail"
                  class="css-1v0ofof-Img e2i8k6n4"
                  height="3"
                  src="https://picsum.photos/id/1024/92/92"
                  width="3"
                />
              </div>
              <div
                class="css-1hg49r7-Box-Flex e2i8k6n2"
                width="100%"
              >
                <h2
                  class="css-x4oqgn-Box e2i8k6n3"
                  font-size="4"
                  font-weight="normal"
                  width="100%"
                >
                  Ocean Measurements
                </h2>
              </div>
            </div>
            <div
              class="css-d9aoz6-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-5jl3ge-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Dataset Description
              </h3>
              <div
                class="css-10zvdr1-Box e2i8k6n3"
                font-size="1"
              >
                Detailed measurements of many ocean locations, including LAT/LON, temperature, humidity and precipitation.
              </div>
            </div>
            <div
              class="css-1rdjp9o-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-1mp3576-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Included countries
              </h3>
              <div
                class="css-10zvdr1-Box e2i8k6n3"
                font-size="1"
              >
                United Kingdom & United States
              </div>
            </div>
            <div
              class="css-14cjmhn-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-1mp3576-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Cost Per Record
              </h3>
              <div
                class="css-1x3sxzl-Box e2i8k6n3"
              >
                $
                0.01
              </div>
            </div>
            <div
              class="css-1ezxmef-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-1mp3576-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Available Records
              </h3>
              <div
                class="css-1x3sxzl-Box e2i8k6n3"
              >
                2250 records
              </div>
            </div>
          </div>
        </div>
        <div>
          <div
            class="css-1y0qilw-Box-Flex e2i8k6n2"
            width="5"
          >
            <div
              class="css-wxi0h7-Box-Flex e2i8k6n2"
            >
              <div
                class="css-z7haim-Box e2i8k6n3"
              >
                <img
                  alt="dataset thumbnail"
                  class="css-1v0ofof-Img e2i8k6n4"
                  height="3"
                  src="https://picsum.photos/id/135/92/92"
                  width="3"
                />
              </div>
              <div
                class="css-1hg49r7-Box-Flex e2i8k6n2"
                width="100%"
              >
                <h2
                  class="css-x4oqgn-Box e2i8k6n3"
                  font-size="4"
                  font-weight="normal"
                  width="100%"
                >
                  Weather and Climate
                </h2>
              </div>
            </div>
            <div
              class="css-d9aoz6-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-5jl3ge-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Dataset Description
              </h3>
              <div
                class="css-10zvdr1-Box e2i8k6n3"
                font-size="1"
              >
                Historical averages and trends of oceanographic climate.
              </div>
            </div>
            <div
              class="css-1rdjp9o-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-1mp3576-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Included countries
              </h3>
              <div
                class="css-10zvdr1-Box e2i8k6n3"
                font-size="1"
              >
                United Kingdom & United States
              </div>
            </div>
            <div
              class="css-14cjmhn-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-1mp3576-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Cost Per Record
              </h3>
              <div
                class="css-1x3sxzl-Box e2i8k6n3"
              >
                $
                0.04
              </div>
            </div>
            <div
              class="css-1ezxmef-Box-Flex e2i8k6n2"
            >
              <h3
                class="css-1mp3576-Box e2i8k6n3"
                color="muted"
                font-size="1"
                font-weight="normal"
                text-decoration="underline"
              >
                Available Records
              </h3>
              <div
                class="css-1x3sxzl-Box e2i8k6n3"
              >
                550 records
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
